<a href="https://colab.research.google.com/github/othniel77/datascience/blob/main/joined_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

In [57]:
from google.colab import files
data_to_load = files.upload()

Saving actors_movies.csv to actors_movies.csv


In [5]:
taxi_owners = pd.read_pickle('/content/taxi_owners.p')
taxi_owners.head()

rid   vid           owner                 address    zip
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618

In [6]:
taxi_veh = pd.read_pickle('/content/taxi_vehicles.p')
taxi_veh.head()

vid    make   model  year fuel_type                owner
0  2767  TOYOTA   CAMRY  2013    HYBRID       SEYED M. BADRI
1  1411  TOYOTA    RAV4  2017    HYBRID          DESZY CORP.
2  6500  NISSAN  SENTRA  2019  GASOLINE       AGAPH CAB CORP
3  2746  TOYOTA   CAMRY  2013    HYBRID  MIDWEST CAB CO, INC
4  5922  TOYOTA   CAMRY  2013    HYBRID       SUMETTI CAB CO

In [7]:
taxi_owners.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3519 entries, 0 to 3518
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   rid      3519 non-null   object
 1   vid      3519 non-null   object
 2   owner    3519 non-null   object
 3   address  3519 non-null   object
 4   zip      3519 non-null   object
dtypes: object(5)
memory usage: 137.6+ KB


In [8]:
taxi_veh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3519 entries, 0 to 3518
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   vid        3519 non-null   object
 1   make       3519 non-null   object
 2   model      3519 non-null   object
 3   year       3519 non-null   int64 
 4   fuel_type  3519 non-null   object
 5   owner      3519 non-null   object
dtypes: int64(1), object(5)
memory usage: 165.1+ KB


In [9]:
# Merge the taxi_owners and taxi_veh tables
taxi_own_veh = taxi_owners.merge(taxi_veh,on='vid')

# Print the column names of the taxi_own_veh
print(taxi_own_veh.columns)

Index(['rid', 'vid', 'owner_x', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_y'],
      dtype='object')


In [10]:
# Merge the taxi_owners and taxi_veh tables setting a suffix
taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid', suffixes =('_own','_veh'))

# Print the column names of taxi_own_veh
print(taxi_own_veh.columns)

Index(['rid', 'vid', 'owner_own', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_veh'],
      dtype='object')


In [11]:

# Print the value_counts to find the most popular fuel_type
print(taxi_own_veh['fuel_type'].value_counts())

HYBRID                    2792
GASOLINE                   611
FLEX FUEL                   89
COMPRESSED NATURAL GAS      27
Name: fuel_type, dtype: int64


 You correctly merged the two tables together and found out that the most common fuel type for taxis in Chicago are hybrids.

## 1.2.Fusion des relations one-to-many (un à plusieurs)

In [14]:
licenses = pd.read_pickle('/content/licenses.p')
licenses.head()

account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613

In [15]:
biz_owners = pd.read_pickle('/content/business_owners.p')
biz_owners.head()

account first_name  last_name      title
0      10      PEARL    SHERMAN  PRESIDENT
1      10      PEARL    SHERMAN  SECRETARY
2   10002     WALTER     MROZEK    PARTNER
3   10002     CELINA     BYRDAK    PARTNER
4   10005      IRENE  ROSENFELD  PRESIDENT

In [16]:
licenses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   account   10000 non-null  object
 1   ward      10000 non-null  object
 2   aid       4657 non-null   object
 3   business  10000 non-null  object
 4   address   10000 non-null  object
 5   zip       9996 non-null   object
dtypes: object(6)
memory usage: 468.9+ KB


In [17]:
biz_owners.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21352 entries, 0 to 21351
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   account     21352 non-null  object
 1   first_name  20721 non-null  object
 2   last_name   20695 non-null  object
 3   title       21341 non-null  object
dtypes: object(4)
memory usage: 667.4+ KB


In [18]:
# Merge the licenses and biz_owners table on account
licenses_owners = licenses.merge(biz_owners,on ='account')

# Group the results by title then count the number of accounts
counted_df = licenses_owners.groupby("title").agg({'account':'count'})

# Sort the counted_df in desending order
sorted_df = counted_df.sort_values("account",ascending=False)

# Use .head() method to print the first few rows of sorted_df
print(sorted_df.head())

                 account
title                   
PRESIDENT           6259
SECRETARY           5205
SOLE PROPRIETOR     1658
OTHER               1200
VICE PRESIDENT       970


You see that president, followed by secretary, are the most common business owner titles.

## 1.3.Fusion de plusieurs tables

In [21]:
ridership=pd.read_pickle('/content/cta_ridership.p')
ridership.head()

station_id  year  month  day  rides
0      40010  2019      1    1    576
1      40010  2019      1    2   1457
2      40010  2019      1    3   1543
3      40010  2019      1    4   1621
4      40010  2019      1    5    719

In [22]:
cal =pd.read_pickle('/content/cta_calendar.p')
cal.head()

year  month  day        day_type
0  2019      1    1  Sunday/Holiday
1  2019      1    2         Weekday
2  2019      1    3         Weekday
3  2019      1    4         Weekday
4  2019      1    5        Saturday

In [23]:
ridership.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3285 entries, 0 to 3284
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   station_id  3285 non-null   object
 1   year        3285 non-null   int64 
 2   month       3285 non-null   int64 
 3   day         3285 non-null   int64 
 4   rides       3285 non-null   int64 
dtypes: int64(4), object(1)
memory usage: 128.4+ KB


In [24]:
cal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   year      365 non-null    int64 
 1   month     365 non-null    int64 
 2   day       365 non-null    int64 
 3   day_type  365 non-null    object
dtypes: int64(3), object(1)
memory usage: 11.5+ KB


In [27]:
stations = pd.read_pickle('/content/stations.p')
stations.head()

station_id        station_name                 location
0      40010  Austin-Forest Park  (41.870851, -87.776812)
1      40020         Harlem-Lake  (41.886848, -87.803176)
2      40030        Pulaski-Lake  (41.885412, -87.725404)
3      40040        Quincy/Wells   (41.878723, -87.63374)
4      40050               Davis   (42.04771, -87.683543)

In [28]:
stations.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 144 entries, 0 to 143
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   station_id    144 non-null    object
 1   station_name  144 non-null    object
 2   location      144 non-null    object
dtypes: object(3)
memory usage: 4.5+ KB


In [29]:
# Merge the ridership and cal tables
ridership_cal = ridership.merge(cal)

In [30]:
# Merge the ridership, cal, and stations tables
ridership_cal_stations = ridership.merge(cal, on=['year','month','day']).merge(stations)

In [31]:
# Merge the ridership, cal, and stations tables
ridership_cal_stations = ridership.merge(cal, on=['year','month','day']).merge(stations, on='station_id')

# Create a filter to filter ridership_cal_stations
filter_criteria = ((ridership_cal_stations['month'] == 7)
                   & (ridership_cal_stations['day_type'] == "Weekday")
                   & (ridership_cal_stations['station_name'] == "Wilson"))

# Use .loc and the filter to select for rides
print(ridership_cal_stations.loc[filter_criteria, 'rides'].sum())

140005


Finally, you found out that the Wilson station had 140,005 riders during weekdays in July.

## 1.4. Fusion de trois tables

In [33]:
zip_demo = pd.read_pickle('/content/zip_demo.p')
zip_demo.head()

zip  income
0  60630   70122
1  60640   50488
2  60622   87143
3  60614  100116
4  60608   41226

In [34]:
zip_demo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66 entries, 0 to 65
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   zip     66 non-null     object
 1   income  66 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.2+ KB


In [38]:
wards = pd.read_pickle('/content/ward.p')
wards.head()

ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649

In [39]:
wards.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   ward      50 non-null     object
 1   alderman  50 non-null     object
 2   address   50 non-null     object
 3   zip       50 non-null     object
dtypes: object(4)
memory usage: 1.7+ KB


In [40]:
# Merge licenses and zip_demo, on zip; and merge the wards on ward
licenses_zip_ward = licenses.merge(zip_demo,on='zip').merge(wards,on ='ward')

# Print the results by alderman and show median income
print(licenses_zip_ward.groupby("alderman").agg({'income':'median'}))

                             income
alderman                           
Ameya Pawar                 66246.0
Anthony A. Beale            38206.0
Anthony V. Napolitano       82226.0
Ariel E. Reyboras           41307.0
Brendan Reilly             110215.0
Brian Hopkins               87143.0
Carlos Ramirez-Rosa         66246.0
Carrie M. Austin            38206.0
Chris Taliaferro            55566.0
Daniel "Danny" Solis        41226.0
David H. Moore              33304.0
Deborah Mell                66246.0
Debra L. Silverstein        50554.0
Derrick G. Curtis           65770.0
Edward M. Burke             42335.0
Emma M. Mitts               36283.0
George Cardenas             33959.0
Gilbert Villegas            41307.0
Gregory I. Mitchell         24941.0
Harry Osterman              45442.0
Howard B. Brookins, Jr.     33304.0
James Cappleman             79565.0
Jason C. Ervin              41226.0
Joe Moore                   39163.0
John S. Arena               70122.0
Leslie A. Hairston          

You see that only a few aldermen represent businesses in areas where the median income is greater than $62,000, which is the median income for the state of Illinois.

## 1.5. la relation un à plusieurs pour la fusion de plusieurs tables

In [43]:
census = pd.read_pickle('/content/census.p')
census.head()

ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637

In [44]:
census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   ward      50 non-null     object
 1   pop_2000  50 non-null     int64 
 2   pop_2010  50 non-null     int64 
 3   change    50 non-null     object
 4   address   50 non-null     object
 5   zip       50 non-null     object
dtypes: int64(2), object(4)
memory usage: 2.5+ KB


In [47]:
land_use = pd.read_pickle('/content/land_use.p')
land_use.head()

ward  residential  commercial  industrial  vacant  other
0    1           41           9           2       2     46
1    2           31          11           6       2     50
2    3           20           5           3      13     59
3    4           22          13           0       7     58
4    5           25           3           1       3     68

In [48]:
land_use.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   ward         50 non-null     object
 1   residential  50 non-null     int64 
 2   commercial   50 non-null     int64 
 3   industrial   50 non-null     int64 
 4   vacant       50 non-null     int64 
 5   other        50 non-null     int64 
dtypes: int64(5), object(1)
memory usage: 2.5+ KB


In [49]:
# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward').merge(licenses, on='ward', suffixes=('_cen','_lic'))

In [50]:
# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward').merge(licenses, on='ward', suffixes=('_cen','_lic'))

# Group by ward, pop_2010, and vacant, then count the # of accounts
pop_vac_lic = land_cen_lic.groupby(["ward","pop_2010","vacant"],
                                   as_index=False).agg({'account':'count'})

Sort pop_vac_lic by vacant, account, and  pop_2010 in descending, ascending, and ascending order respectively. Save it as sorted_pop_vac_lic.

In [51]:

# Sort pop_vac_lic and print the results
sorted_pop_vac_lic = pop_vac_lic.sort_values(["vacant","account","pop_2010"],
                                             ascending=(False,True,True))

# Print the top few rows of sorted_pop_vac_lic
print(sorted_pop_vac_lic.head())

   ward  pop_2010  vacant  account
47    7     51581      19       80
12   20     52372      15      123
1    10     51535      14      130
16   24     54909      13       98
7    16     51954      13      156


you were able to pull together information from different tables to see that the 7th ward would be a good place to build your goat farm!

# 2. Fusion avec les différentes types de jointures

### 2.1. la jointure Left join

In [59]:
toy_story = pd.read_pickle('/content/movie_to_genres.p')
toy_story.head()

movie_id            genre
0         5            Crime
1         5           Comedy
2        11  Science Fiction
3        11           Action
4        11        Adventure

In [ ]:
# Merge the toy_story and taglines tables with a left join
toystory_tag = toy_story.merge(taglines,on='id',how='left')

# Print the rows and shape of toystory_tag
print(toystory_tag)
print(toystory_tag.shape)